Expose additional properties during cluster creation

brianhealeyRMN opened the pull request:
Allow for sending IssueClientCertificate as part of the cluster creation request
Allow for sending the Management object as part of the cluster creation request
Allow for sending the AutoScale object as part of the cluster creation request

danawillow commented on google/resource_container_cluster.go:
```diff
@@ -266,6 +271,60 @@ func resourceContainerCluster() *schema.Resource {
+							},
+						},
+
+						"auto_scaling": {
```
Can this be `autoscaling` to match what's in `google_container_node_pool` and the API docs (which has it as one word)?

brianhealeyRMN replied:
Modified as suggested

danawillow commented on google/resource_container_cluster.go:
```diff
@@ -476,13 +587,23 @@ func resourceContainerClusterRead(d *schema.ResourceData, meta interface{}) erro
+			"client_certificate":       cluster.MasterAuth.ClientCertificate,
+			"client_key":               cluster.MasterAuth.ClientKey,
+			"cluster_ca_certificate":   cluster.MasterAuth.ClusterCaCertificate,
+			"issue_client_certificate": true,
```
Instead of setting it to true here and potentially overriding it later, how about just set it later?

brianhealeyRMN replied:
Not a problem. Default behavior is "true". This is a flag vs an actual setting so I added code so that it behaves "as a setting" by checking for the existence of the client certitiface

danawillow commented on google/resource_container_cluster_test.go:
```diff
@@ -635,6 +687,7 @@ resource "google_container_cluster" "with_master_auth" {
 	master_auth {
 		username = "mr.yoda"
 		password = "adoy.rm"
+		issue_client_certificate = false
```
Does this change anything from what was there before? I think if you want to test `issue_client_certificate` it probably makes more sense to set it to `true`

brianhealeyRMN replied:
True is the default behavior, so setting it to false does change, however I think the larger concern is that both cases were not being tested.  I added test specifically to test true/false

danawillow commented on google/resource_container_cluster.go:
```diff
@@ -66,6 +66,11 @@ func resourceContainerCluster() *schema.Resource {
 						},
+						"issue_client_certificate": {
+							Type:             schema.TypeBool,
+							Optional:         true,
```
I think this needs to be `ForceNew` as well.

brianhealeyRMN replied:
Done

brianhealeyRMN commented on google/resource_container_cluster.go:
```diff
@@ -66,6 +66,11 @@ func resourceContainerCluster() *schema.Resource {
 							Required: true,
 							ForceNew: true,
 						},
+						"issue_client_certificate": {
```
Updated the orders for the schema to be alphabetical

brianhealeyRMN commented on google/resource_container_cluster.go:
```diff
@@ -66,6 +66,11 @@ func resourceContainerCluster() *schema.Resource {
+						"issue_client_certificate": {
+							Type:             schema.TypeBool,
+							Optional:         true,
+							DiffSuppressFunc: linkDiffSuppress,
```
Removed

brianhealeyRMN commented on google/resource_container_cluster.go:
```diff
@@ -266,6 +271,60 @@ func resourceContainerCluster() *schema.Resource {
 
 						"node_config": schemaNodeConfig,
+
+						"management": {
```
done

brianhealeyRMN commented on google/resource_container_cluster.go:
```diff
@@ -266,6 +271,60 @@ func resourceContainerCluster() *schema.Resource {
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+
+									"enabled": {
```
Modified as suggested

brianhealeyRMN commented:
Thanks @danawillow 
I updated the the PR with your suggestions and fixed the integration tests.  I also added additional tests to provide better coverage.

davidquarles commented:
Yeah, this is awesome @brianhealeyRMN!  @danawillow and I were discussing version upgrades in #633, which led me here.  Would it be possible to extend the `nodePool.Management` work to target the node pools directly, rather than (or in addition to?) the parent cluster resource, and to handle the update lifecycle as well?  I can totally help, if you'd like to collaborate.  I can also open a new issue and start from scratch, if this is entirely out of scope here.

justinsb commented:
I would love to see the `issue_client_certificate` support land, that was implemented in this PR.  @brianhealeyRMN do you think you might rebase or split this PR up?  Or would you mind if I (or someone else, if anyone else is keen) added that support?

danawillow commented:
I'm not @brianhealeyRMN, but I think you should go for it @justinsb :)

paddycarver commented:
This PR has been inactive for a long time, with no response, so I'm going to close it out. If the author wants to carry on with it, feel free to post in this thread, and we can re-open it. If someone else is interested in picking this up, feel free to open a new PR for it.
